packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_connection.dart

```diff
@@ -36,34 +36,86 @@ class ChartConnection extends DisposableController
   );
 
   Timer? _pollingTimer;
-  bool _connected = false;
 
-  late final isDeviceAndroid =
-      serviceConnection.serviceManager.vm?.operatingSystem == 'android';
+  /// If completed, this instance was connected to the application.
+  final Completer<void> _initialized = Completer();
+
+  void _stopConnection() {
+    _pollingTimer?.cancel();
+    cancelStreamSubscriptions();
+    cancelListeners();
+  }
+
+  // True if connection was started and then stopped.
+  bool get _isConnectionStopped {
+    assert(_initialized.isCompleted);
+    return _pollingTimer?.isActive == false;
+  }
+
+  late bool isDeviceAndroid;
+
+  /// True if DevTools is in connected mode.
+  ///
+  /// If DevTools is in offline mode, stops connection and returns false.
+  bool _checkConnection() {
+    assert(_initialized.isCompleted);
+    if (_isConnectionStopped) return false;
+
+    // If connection is up and running, return true.
+    if (!offlineDataController.showingOfflineData.value &&
+        serviceConnection.serviceManager.connectedState.value.connected) {
+      return true;
+    }
+
+    // Otherwise stop connection and return false.
+    _stopConnection();
+    return false;
+  }
+
+  Future<void> maybeInitialize() async {
+    if (_initialized.isCompleted) return;
+    _initialized.complete();
+    if (!_checkConnection()) return;
 
-  Future<void> maybeConnect() async {
-    if (_connected) return;
     await serviceConnection.serviceManager.onServiceAvailable;
+
+    isDeviceAndroid =
+        serviceConnection.serviceManager.vm?.operatingSystem == 'android';
+
+    addAutoDisposeListener(
+      serviceConnection.serviceManager.connectedState,
+      _checkConnection,
+    );
+
     autoDisposeStreamSubscription(
       serviceConnection.serviceManager.service!.onExtensionEvent
           .listen(_memoryTracker.onMemoryData),
     );
+
     autoDisposeStreamSubscription(
       serviceConnection.serviceManager.service!.onGCEvent
           .listen(_memoryTracker.onGCEvent),
     );
-    await _onPoll();
-    _connected = true;
+
+    await _startPolling();
   }
 
-  Future<void> _onPoll() async {
-    _pollingTimer = null;
-    await _memoryTracker.pollMemory();
-    _pollingTimer = Timer(chartUpdateDelay, _onPoll);
+  Future<void> _startPolling() async {
+    assert(_initialized.isCompleted);
+    if (!_checkConnection()) return;
+    try {
+      await _memoryTracker.pollMemory();
+      // Timer is not periodic because we do not want polls to overlap.
+      _pollingTimer = Timer(chartUpdateDelay, _startPolling);
+    } catch (e) {
+      if (_checkConnection()) rethrow;
+    }
   }
 
   @override
   void dispose() {
+    // Not nulling out timer, because we need timer to be not null and inactive
+    // for `_isConnectionStopped` to return true.
```

Comment:
this is not actually true. `_isConnectionStopped` handles a null `_pollingTimer` object as intended, so we should null it out here.

Reply:
maybeInitialize has async gap when _initialized is already completed, but _polling is still null. 
I updated _isConnectionStopped

Observation: if we had a flag `isDisconnected` we would not spend this much time discussing this question.